### PR TITLE
WIP: make tests accept a list of tests to be run

### PR DESCRIPTION
**Pull request by mmacata**
For running all tests with multirunner wrapped in testsuite, I like to control which tests to run. Therefore this pull request adds arguments which
- allow to read from a list on filesystem to control which tests to run
- allow to write all testmodules to a list on filesystem, which can help to create above list

This PR is WIP (Work in Progress) as it currently fails if READ_MODULE_TEST_LIST or WRITE_MODULE_TEST_LIST are set to "no" because it passes a NoneType argument to a subprocess in multirunner.py.

Comments welcome :)

### CODE REVIEW

File: lib/python/gunittest/invoker.py [hunk 1]
```diff
@@ -231,10 +231,29 @@ def run_in_location(self, gisdbase, location, location_type,
...
+                with open(test_list_out, 'a') as file:
+                    file.write(module.abs_file_path + "\n")
+
         self.reporter.start(results_dir)
+
         for module in modules:
-            self._run_test_module(module=module, results_dir=results_dir,
-                                  gisdbase=gisdbase, location=location)
+            # TODO: check if it can be None here or breaks before
+            if test_list_in:
```
Comment (pmav99) [inline lib/python/gunittest/invoker.py]: Doesn't this mean that the tests will run if and only if `test_list_in` has been provided?

File: lib/python/gunittest/invoker.py [hunk 2]
```diff
@@ -206,7 +206,7 @@ def _run_test_module(self, module, results_dir, gisdbase, location):
             shutil.rmtree(mapset_dir)
 
     def run_in_location(self, gisdbase, location, location_type,
-                        results_dir):
+                        results_dir, test_list_in, test_list_out):
```
Comment (pmav99) [inline lib/python/gunittest/invoker.py]: If anyone is calling `run_in_location()`, their code will be broken after this commit since they will only be giving 4 arguments instead of 6. Perhaps `test_list_in` and `test_list_out` could have some default values?

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf [hunk 3]
```diff
@@ -27,3 +27,15 @@ REPORTS="testreports"
 PUBLISH="no"
 # upload WWW dir on server for report publication (not used for local tests)
 SERVERDIR="/var/www/html/grassgistestreports"
+
+# read module tests list from file
+# may be no|yes
+READ_MODULE_TEST_LIST=yes
```
Comment (pmav99) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: Is it wise to default this to `yes`? 
I think not unless you also commit the `test.list` file.
Of course, the contents of this file will need to be the full list of existing tests (and this list will need to be maintained too).

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf [hunk 4]
```diff
@@ -27,3 +27,15 @@ REPORTS="testreports"
...
+# file with list of tests to be run
+TEST_LIST_IN=/src/grass_build/testsuite/examples/test.list
```
Comment (pmav99) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: I would prefer if paths were not hardcoded. See also #5 
You could probably use something like: `"${GRASSSRC}/path/to/test.list"`

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf [hunk 5]
```diff
@@ -27,3 +27,15 @@ REPORTS="testreports"
...
+
+# write module tests list to file (help to prepare above list)
+# may be no|yes
+WRITE_MODULE_TEST_LIST=yes
+# file in which the list should be written
+TEST_LIST_OUT=/src/grass_build/testsuite/examples/tests_complete.list
```
Comment (pmav99) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: I would prefer if paths were not hardcoded. See also #5
You could probably use something like: `"${GRASSSRC}/path/to/test_complete.list"`

**Review by pmav99 — CHANGES_REQUESTED** [review]
Running only part of the test-suite is something that is really needed, but IMHO before merging there are some details that need to be taken care of.

File: lib/python/gunittest/main.py [hunk 6]
```diff
@@ -135,6 +135,11 @@ def main():
     parser.add_argument('--output', dest='output', action='store',
                         default='testreport',
                         help='Output directory')
+    parser.add_argument('--test_list_in',
```
Comment (pmav99) [inline lib/python/gunittest/main.py:138]: here you are adding two mandatory arguments to a CLI utility. 
I am not saying that breaking APIs is necessarily a problem (for internal APIs it is usually not)  but external APIs are a bit more sensitive.
Maybe adding some default values would make sense :)

**Comment by pmav99** [comment]
I think it would be a good idea, before making API breaking changes to ask on the dev list if people use these test-runners and how. E.g. if most people only care/want to run the full test-suite, then keeping this option as the default and making piece-wise selection of tests optional would probably be preferred.

**Comment by mmacata** [comment]
Thank you for your comments! You are right, I didn't want to merge like this but still working on it (WIP)  to already get some feedback. The goal is to not make API breaking changes at all and to have an expert looking over it to see where things might break. Thank you, I will keep working on it

File: lib/python/gunittest/invoker.py (hunk 1 above)
Comment (mmacata) [reply lib/python/gunittest/invoker.py]: Good point, I changed it

File: lib/python/gunittest/invoker.py (hunk 2 above)
Comment (mmacata) [reply lib/python/gunittest/invoker.py]: True, now they have default values.

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf (hunk 3 above)
Comment (mmacata) [reply testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: Absolutely right, I changed it to 'no'

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf (hunk 4 above)
Comment (mmacata) [reply testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: changed this one as well

File: lib/python/gunittest/main.py (hunk 6 above)
Comment (mmacata) [reply lib/python/gunittest/main.py:138]: I also prefer not breaking APIs if possible, so I changed it to use default values

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf [hunk 7]
```diff
@@ -30,12 +30,12 @@ SERVERDIR="/var/www/html/grassgistestreports"
 
 # read module tests list from file
 # may be no|yes
-READ_MODULE_TEST_LIST=yes
+READ_MODULE_TEST_LIST=no
 # file with list of tests to be run
-TEST_LIST_IN=/src/grass_build/testsuite/examples/test.list
+TEST_LIST_IN="${GRASSSRC}/path/to/test.list"
```
Comment (pmav99) [inline testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: Does the `/path/to` directory exist in the repo? How will our code behave when it tries to access it? Let's assume that someone only sets `READ_MODULE_TEST_LIST` to `yes`, without changing the path, will it work?

The same is true for `TEST_LIST_OUT` too.

File: lib/python/gunittest/invoker.py [hunk 8]
```diff
@@ -206,7 +206,8 @@ def _run_test_module(self, module, results_dir, gisdbase, location):
             shutil.rmtree(mapset_dir)
 
     def run_in_location(self, gisdbase, location, location_type,
-                        results_dir, test_list_in, test_list_out):
+                        results_dir, test_list_in='False',
+                        test_list_out='False'):
```
Comment (pmav99) [inline lib/python/gunittest/invoker.py:244]: Using a string containing the value `False` will almost certainly cause some head scratching to the average Python developer. Too easy to confuse it with the boolean `False` while at the same time, the comparison semantics are different. Please check this:
``` python
value = 'False'
if value:
    print('str: will I be printed?')
# vs
value = False
if value:
    print('bool: will I be printed?')
```
and read the following:
- https://docs.python-guide.org/writing/style/#check-if-a-variable-equals-a-constant
- https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce
- https://stackoverflow.com/questions/2988017/string-comparison-in-python-is-vs

good knowledge of comparison semantics is essential.

Anyway, in order to help you move forward, I believe that the correct approach in this case would be to use `None` as the default value. I.e. `test_list_out` is either a string containing a path to the output file or `None` if the output file was not specified. 
Note: this will require that you also change the comparisons that follow (e.g. `if test_list_out != 'False'` etc) + the defaults arguments for `argparse`.

**Review by pmav99 — CHANGES_REQUESTED** [review]
That's probably better, but I am afraid still not quite there yet... Please see the comments)

File: testsuite/examples/test_framework_GRASS_GIS_with_NC.conf (hunk 7 above)
Comment (mmacata) [reply testsuite/examples/test_framework_GRASS_GIS_with_NC.conf]: No it doesn't exist. And if someone sets `READ_MODULE_TEST_LIST `to `yes` it will not work. But would't the same apply to line 20 where a configure script is called? As this file is a configuration in an example folder it might not be too bad but I changed it.

File: lib/python/gunittest/invoker.py (hunk 8 above)
Comment (mmacata) [reply lib/python/gunittest/invoker.py:244]: Ok thank you. I initially had it set to `None`. In this case I would still need to test if it is `None` or `"False"` because this is the default value passed from the parser in `lib/python/gunittest/main.py`, passed from `lib/python/gunittest/multirunner.py`. There I can't use `None` because `subprocess.Popen() `would fail then. Do you think it would be better to pass the  `READ_MODULE_TEST_LIST` and `WRITE_MODULE_TEST_LIST` flags from `testsuite/examples/test_framework_GRASS_GIS_with_NC.sh` to the multirunner and then formulate two different calls to run `grass.gunittest.main`? Or should I test if it is `None` or `"False"` in `invoker.py`? Or better rename `"False"` in `multirunner.py` and `main.py` to a word that has no oher meaning in python? Thank you for your help

Comment (pmav99) [reply lib/python/gunittest/invoker.py:244]: Run [this](https://gist.github.com/pmav99/880d7e4de5d04098417fdecffbee9dea) with the following arguments:
```
python foo.py                                # no arguments -> i.e. you use the defaults
python foo.py --mypath ''                    # specify the empty string
python foo.py --mypath '/path/to/a/file'     # specify a proper path.
```
Try to understand what happends with the boolean values.

Hopefully, you will find it helpful. If not let me know.

**Comment by pmav99** [comment]
@neteler although the debate for merging vs rebasing is somewhat philosophical, for feature branches rebasing is usually the preferred option since it leads to cleaner history.

**Comment by neteler** [comment]
> @neteler although the debate for merging vs rebasing is somewhat philosophical, for feature branches rebasing is usually the preferred option since it leads to cleaner history.

ok, I probably got confused with GH functionality here...

**Comment by neteler** [comment]
@mmacata can you please move this PR over to https://github.com/OSGeo/grass/pulls ? That is the new repository (this here will be deleted as soon as your PR has been transferred).

Thanks!
